### ISSUE TITLE
Slower performance in GPU than in CPU

### ISSUE THREAD

**Issue by pythondlml** [plaidml/plaidml#333]
Hi,
I am using plaidML to run some experiments using Keras examples.
But for some cases I tested, it turns out it takes longer time using GPU than using CPU.
Here is the specs of the machine.
CPU: 2.7 GHz Intel Core i7
OS: macOS 10.12.16
INFO:plaidml:Opening device "metal_amd_radeon_pro_460.0"

Example 1: train word embedding on IMDB data set 
https://github.com/fchollet/deep-learning-with-python-notebooks/blob/master/6.1-using-word-embeddings.ipynb
- using GPU: 5 s/epoch
- using CPU: 1 s/epoch
- codes:
`model = Sequential()`
`model.add(Embedding(10000, 8, input_length=20))`
`model.add(Flatten())`
`model.add(Dense(1, activation='sigmoid'))`
`model.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['acc'])`

Example 2: train word embedding and GRU on IMDB data set
- using GPU: 65 s/epoch
- using CPU: 11 s/epoch
- codes:
`model = Sequential()`
`model.add(layers.Embedding(10000, 16, input_length=100))`
`model.add(layers.Conv1D(16, 7, activation='relu'))`
`model.add(layers.MaxPooling1D(3))`
`model.add(layers.Conv1D(16, 7, activation='relu'))`
`model.add(layers.GRU(16, dropout=0.1, recurrent_dropout=0.5))`
`model.add(layers.Dense(1))`

But I did find the case where using GPU is faster than using CPU.
Example 3: train convnet on the MNIST
https://github.com/keras-team/keras/blob/master/examples/mnist_cnn.py
- using GPU: 27 s/epoch
- using CPU: 97 s/epoch

I found some discussion treads about training LSTM or embedding is slower in GPU than in CPU with plaidML.
https://github.com/plaidml/plaidml/issues/171
https://github.com/plaidml/plaidml/issues/312

Is this the reason why I see a slower performance in example 1 and example 2? If so, any idea how soon this might be resolved?

Any thought is appreciated. 
Thanks,

**Comment by dgkutnic** [comment]
Thanks for providing us with these examples! I'm going to reproduce them on my end to see if they can be traced back to the embedding issue.

If this slowness is coming from the embedding issue, please note that while it's on our radar, it is a major revision on our end. We're actively working on this and these additional examples are extremely helpful with our debug and for moving this up in our priority list.

**Comment by tux-o-matic** [comment]
@pythondlml , what about batch size as mention in #50? GPU and CPU don't offer the same capacity to load data, here with plaidML more than anything it seems one must tune the batch size.

**Comment by dgkutnic** [comment]
We've been able to isolate the slowness to our implementation of scatter/gather, which are used heavily in the Embedding layers. Across multiple batch sizes and sizes of the embedding layer, we have consistently seen the slowness issue that @pythondlml has reported.

**Comment by brianretford** [comment]
Duplicate of #96
